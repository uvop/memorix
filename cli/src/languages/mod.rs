mod python;
mod rust;
mod typescript;

use crate::{
    flat_schema::FlatValidatedSchema,
    parser::{Export, Language},
    validate::ValidatedSchema,
};
use std::collections::HashMap;

pub fn codegen_per_language(
    export: &Export,
    validated_schema: &ValidatedSchema,
    flat_validated_schema: &FlatValidatedSchema,
) -> HashMap<String, String> {
    export
        .files
        .as_ref()
        .unwrap_or(&vec![])
        .iter()
        .map(|file| {
            let path = file.path.to_string();
            let content = match file.language {
                Language::TypeScript => typescript::codegen(validated_schema),
                Language::Python => python::codegen(flat_validated_schema),
                Language::Rust => rust::codegen(flat_validated_schema),
            };
            (path, content)
        })
        .collect()
}
