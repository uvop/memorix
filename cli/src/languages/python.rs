use crate::{
    flat_schema::{
        FlatValidatedNamespace, FlatValidatedReferenceTypeItemKind, FlatValidatedSchema,
        FlatValidatedTypeItem, TypeObjectItem,
    },
    parser::{Engine, Value, ALL_CACHE_OPERATIONS, ALL_PUBSUB_OPERATIONS, ALL_TASK_OPERATIONS},
};

fn indent(level: usize) -> String {
    "    ".repeat(level)
}
fn flat_type_item_to_code(
    flat_type_item: &FlatValidatedTypeItem,
    schema: &FlatValidatedSchema,
) -> String {
    match flat_type_item {
        FlatValidatedTypeItem::Optional(x) => {
            format!("typing.Optional[{}]", flat_type_item_to_code(x, schema)).to_string()
        }
        FlatValidatedTypeItem::Array(x) => {
            format!("typing.List[{}]", flat_type_item_to_code(x, schema)).to_string()
        }
        FlatValidatedTypeItem::Reference(x) => {
            let (namespace, namespace_names) = x.namespace_indexes.iter().fold(
                (&schema.global_namespace, Vec::<&str>::new()),
                |(current_namespace, mut acc), &i| {
                    let (next_name, next_namespace) = &current_namespace.namespaces[i];
                    acc.push(next_name.as_str());
                    (next_namespace, acc)
                },
            );
            format!(
                "{prefix}{seperator}{name}",
                prefix = namespace_names
                    .iter()
                    .map(|x| stringcase::pascal_case(x))
                    .collect::<Vec<_>>()
                    .join("."),
                seperator = match namespace_names.is_empty() {
                    true => "",
                    false => ".",
                },
                name = match x.kind {
                    FlatValidatedReferenceTypeItemKind::TypeItem(i) => {
                        namespace.flat_type_items[i].0.clone()
                    }
                    FlatValidatedReferenceTypeItemKind::TypeObjectItem(i) => {
                        namespace.type_object_items[i].0.clone()
                    }
                    FlatValidatedReferenceTypeItemKind::Enum(i) => {
                        namespace.enum_items[i].0.clone()
                    }
                }
            )
        }
        FlatValidatedTypeItem::Boolean => "bool".to_string(),
        FlatValidatedTypeItem::String => "str".to_string(),
        FlatValidatedTypeItem::U32 => "int".to_string(),
        FlatValidatedTypeItem::U64 => "int".to_string(),
        FlatValidatedTypeItem::I32 => "int".to_string(),
        FlatValidatedTypeItem::I64 => "int".to_string(),
        FlatValidatedTypeItem::F32 => "float".to_string(),
        FlatValidatedTypeItem::F64 => "float".to_string(),
    }
}

fn value_to_code(value: &Value) -> String {
    match value {
        Value::String(x) => format!("\"{x}\""),
        Value::Env(x) => {
            format!("os.environ[\"{x}\"]")
        }
    }
}

fn type_item_object_to_code(
    name: &str,
    type_item_object: &TypeObjectItem,
    schema: &FlatValidatedSchema,
    level: usize,
) -> String {
    let base_indent = indent(level);
    format!(
        r#"
{base_indent}@dataclass
{base_indent}class {name}(object):
{}
"#,
        type_item_object
            .properties
            .iter()
            .map(|(property_name, flat_type_item)| format!(
                "{base_indent}    {property_name}: {}",
                flat_type_item_to_code(flat_type_item, schema)
            )
            .to_string())
            .collect::<Vec<_>>()
            .join("\n")
    )
    .to_string()
}

fn namespace_to_code(
    namespace: &FlatValidatedNamespace,
    name_tree: Vec<String>,
    schema: &FlatValidatedSchema,
) -> String {
    let mut result = String::from("");
    let base_indent = indent(name_tree.len());

    for (name, values) in &namespace.enum_items {
        result.push_str(&format!(
            r#"
{base_indent}class {name}(str, Enum):
{}

"#,
            values
                .iter()
                .map(|x| format!("{base_indent}    {x} = \"{x}\""))
                .collect::<Vec<_>>()
                .join("\n")
        ));
    }
    for (name, type_item_object) in &namespace.type_object_items {
        result.push_str(&format!(
            "{}\n",
            type_item_object_to_code(name.as_str(), type_item_object, schema, name_tree.len())
        ));
    }
    for (name, flat_type_item) in &namespace.flat_type_items {
        result.push_str(&format!(
            "{base_indent}{name} = {}\n\n",
            flat_type_item_to_code(flat_type_item, schema)
        ));
    }
    for (name, sub_namespace) in &namespace.namespaces {
        result.push_str(&format!(
            r#"{base_indent}class {name}(object):
{namespace_content}
"#,
            name = stringcase::pascal_case(name),
            namespace_content = namespace_to_code(
                sub_namespace,
                name_tree
                    .clone()
                    .into_iter()
                    .chain(std::iter::once(name.clone()))
                    .collect(),
                schema,
            ),
        ));
    }
    if !namespace.cache_items.is_empty() {
        result.push_str(&format!(
            r#"
{base_indent}class MemorixCache(MemorixCacheAll.Base):
{base_indent}    def __init__(self, api: MemorixBase) -> None:
{base_indent}        super().__init__(api=api)

{class_content}

"#,
            class_content = namespace
                .cache_items
                .iter()
                .map(|(name, item)| {
                    let payload = flat_type_item_to_code(&item.payload, schema);
                    format!(
                        r#"{base_indent}        self.{name} = MemorixCacheAll.Item{api}{key}](
{base_indent}            api=api,
{base_indent}            id="{name}",
{base_indent}            payload_class={payload},{options}
{base_indent}        )"#,
                        key = match &item.key {
                            None => format!("NoKey[{payload}"),
                            Some(key) => {
                                let key = flat_type_item_to_code(key, schema);
                                format!("[{key}, {payload}")
                            }
                        },
                        api = ALL_CACHE_OPERATIONS
                            .iter()
                            .map(|x| match item.expose.contains(x) {
                                true => "T",
                                false => "F",
                            })
                            .collect::<Vec<_>>()
                            .join(""),
                        options = {
                            let content = [
                                item.ttl.as_ref().map(|x| {
                                    format!(
                                        "{base_indent}                ttl={},",
                                        value_to_code(x)
                                    )
                                }),
                                item.extend_on_get.as_ref().map(|x| {
                                    format!(
                                        "{base_indent}                extend_on_get={},",
                                        value_to_code(x)
                                    )
                                }),
                            ]
                            .into_iter()
                            .flatten()
                            .collect::<Vec<_>>()
                            .join("\n");
                            match content.is_empty() {
                                true => "".to_string(),
                                false => format!(
                                    r#"
{base_indent}            options=MemorixCacheAll.Options(
{content}
{base_indent}            )"#
                                ),
                            }
                        }
                    )
                })
                .collect::<Vec<_>>()
                .join("\n"),
        ));
    }
    if !namespace.pubsub_items.is_empty() {
        result.push_str(&format!(
            r#"
{base_indent}class MemorixPubSub(MemorixPubSubAll.Base):
{base_indent}    def __init__(self, api: MemorixBase) -> None:
{base_indent}        super().__init__(api=api)

{class_content}

"#,
            class_content = namespace
                .pubsub_items
                .iter()
                .map(|(name, item)| {
                    let payload = flat_type_item_to_code(&item.payload, schema);
                    format!(
                        r#"{base_indent}        self.{name} = MemorixPubSubAll.Item{api}{key}](
{base_indent}            api=api,
{base_indent}            id="{name}",
{base_indent}            payload_class={payload},
{base_indent}        )"#,
                        key = match &item.key {
                            None => format!("NoKey[{payload}"),
                            Some(key) => {
                                let key = flat_type_item_to_code(key, schema);
                                format!("[{key}, {payload}")
                            }
                        },
                        api = ALL_PUBSUB_OPERATIONS
                            .iter()
                            .map(|x| match item.expose.contains(x) {
                                true => "T",
                                false => "F",
                            })
                            .collect::<Vec<_>>()
                            .join("")
                    )
                })
                .collect::<Vec<_>>()
                .join("\n"),
        ));
    }
    if !namespace.task_items.is_empty() {
        result.push_str(&format!(
            r#"
{base_indent}class MemorixTask(MemorixTaskAll.Base):
{base_indent}    def __init__(self, api: MemorixBase) -> None:
{base_indent}        super().__init__(api=api)

{class_content}

"#,
            class_content = namespace
                .task_items
                .iter()
                .map(|(name, item)| {
                    let payload = flat_type_item_to_code(&item.payload, schema);
                    format!(
                        r#"{base_indent}        self.{name} = MemorixTaskAll.Item{api}{key}](
{base_indent}            api=api,
{base_indent}            id="{name}",
{base_indent}            payload_class={payload},{options}
{base_indent}        )"#,
                        key = match &item.key {
                            None => format!("NoKey[{payload}"),
                            Some(key) => {
                                let key = flat_type_item_to_code(key, schema);
                                format!("[{key}, {payload}")
                            }
                        },
                        api = ALL_TASK_OPERATIONS
                            .iter()
                            .map(|x| match item.expose.contains(x) {
                                true => "T",
                                false => "F",
                            })
                            .collect::<Vec<_>>()
                            .join(""),
                        options = {
                            let content = [item.queue_type.as_ref().map(|x| {
                                format!(
                                    "{base_indent}                queue_type={},",
                                    value_to_code(x)
                                )
                            })]
                            .into_iter()
                            .flatten()
                            .collect::<Vec<_>>()
                            .join("\n");
                            match content.is_empty() {
                                true => "".to_string(),
                                false => format!(
                                    r#"
{base_indent}            options=MemorixTaskAll.Options(
{content}
{base_indent}            )"#
                                ),
                            }
                        }
                    )
                })
                .collect::<Vec<_>>()
                .join("\n"),
        ));
    }
    let pre_namespace = match name_tree.is_empty() {
        true => "".to_string(),
        false => format!(
            "{}.",
            name_tree
                .iter()
                .map(|x| stringcase::pascal_case(x))
                .collect::<Vec<_>>()
                .join(".")
        ),
    };
    result.push_str(&format!(
        r#"{base_indent}class Memorix(MemorixBase):
{init_def}

{base_indent}        self._namespace_name_tree = [{name_tree}]
{class_content}
"#,
        name_tree = name_tree
            .iter()
            .map(|x| format!("\"{}\"", x))
            .collect::<Vec<_>>()
            .join(", "),
        init_def = match name_tree.is_empty() {
            true => format!(
                r#"{base_indent}    def __init__(self) -> None:
{base_indent}        super().__init__(redis_url={redis_url})
"#,
                redis_url = match &schema.engine {
                    Engine::Redis(x) => value_to_code(x),
                }
            ),
            false => format!(
                r#"{base_indent}    def __init__(self, ref: MemorixBase) -> None:
{base_indent}        super().__init__(ref=ref)
"#
            ),
        },
        class_content = [
            namespace
                .namespaces
                .iter()
                .map(|(name, _)| format!(
                    "{base_indent}        self.{name} = {pre_namespace}{namespace_name}.Memorix(ref=self)",
                    namespace_name = stringcase::pascal_case(name)
                ))
                .collect::<Vec<_>>()
                .join("\n"),
            [
                (!namespace.cache_items.is_empty())
                    .then(|| format!("{base_indent}        self.cache = {pre_namespace}MemorixCache(self)")),
                (!namespace.pubsub_items.is_empty())
                    .then(|| format!("{base_indent}        self.pubsub = {pre_namespace}MemorixPubSub(self)")),
                (!namespace.task_items.is_empty())
                    .then(|| format!("{base_indent}        self.task = {pre_namespace}MemorixTask(self)")),
            ]
            .into_iter()
            .flatten()
            .collect::<Vec<_>>()
            .join("\n")
        ]
        .into_iter()
        .filter(|x| !x.is_empty())
        .collect::<Vec<_>>()
        .join("\n\n"),
    ));
    result
}

pub fn codegen(schema: &FlatValidatedSchema) -> String {
    format!(
        r#"# flake8: noqa
import typing
import os

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from enum import Enum
from memorix_client_redis import (
    MemorixBase,
    MemorixCacheAll,
    MemorixPubSubAll,
    MemorixTaskAll,
)

{}"#,
        namespace_to_code(&schema.global_namespace, vec![], schema)
    )
    .to_string()
}
