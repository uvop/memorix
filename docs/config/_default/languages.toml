[en]
  languageName = "English"
  contentDir = "content/en"
  weight = 10
  [en.params]
    languageISO = "EN"
    languageTag = "en-US"

[he]
  languageName = "Hebrew"
  contentDir = "content/he"
  weight = 15
  [de.params]
    languageISO = "HE"
    languageTag = "he"
