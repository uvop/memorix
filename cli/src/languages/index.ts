import { assertUnreachable } from "src/core/utilities";
import { codegenTs } from "./ts/ts";
import { codegenPython } from "./python/python";

export enum Languages {
  typescript = "typescript",
  python = "python",
}

type CodegenByLanguageFn = (schema: string, language: Languages) => string;

export const codegenByLanguage: CodegenByLanguageFn = (scope, language) => {
  switch (language) {
    case Languages.typescript:
      return codegenTs(scope);
    case Languages.python:
      return codegenPython(scope);
    default:
      assertUnreachable(language);
      return "";
  }
};
