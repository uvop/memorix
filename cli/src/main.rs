mod export_schema;
mod flat_schema;
mod imports;
mod languages;
mod parser;
mod parser_tools;
mod validate;

use std::error::Error;
use std::fmt;
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::PathBuf;

use languages::codegen_per_language;
use nom::error::convert_error;
use nom::Finish;
use parser::{schema_to_sdl, Schema};
use parser_tools::FromSdl;

pub struct PrettyError<T>(pub T);

impl<T> From<T> for PrettyError<T> {
    fn from(v: T) -> Self {
        Self(v)
    }
}

impl<T: fmt::Display> fmt::Debug for PrettyError<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(&self.0, f)
    }
}

impl From<Box<dyn Error>> for PrettyError<String> {
    fn from(error: Box<dyn Error>) -> Self {
        PrettyError(error.to_string())
    }
}

impl<T: fmt::Debug + fmt::Display> Error for PrettyError<T> {}

impl<T: fmt::Display> fmt::Display for PrettyError<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

pub trait FileSystem {
    fn read_to_string(&self, path: &str) -> io::Result<String>;
    fn write_string(&self, path: &str, content: &str) -> io::Result<()>;
}

pub fn resolve_path(relative_file: &str, target_file: &str) -> io::Result<String> {
    let parent_path = std::path::Path::new(relative_file)
        .parent()
        .ok_or(io::Error::new(
            io::ErrorKind::NotFound,
            "Couldn't find parent dir".to_string(),
        ))?;

    let path = std::path::Path::new(&target_file);
    Ok(match path.is_absolute() {
        true => target_file.to_string(),
        false => parent_path
            .join(path)
            .to_str()
            .ok_or(io::Error::new(
                io::ErrorKind::NotFound,
                "Couldn't append parent dir and codedir",
            ))?
            .to_string(),
    })
}

pub struct RealFileSystem;

impl FileSystem for RealFileSystem {
    fn read_to_string(&self, path: &str) -> io::Result<String> {
        let mut file = File::open(path)?;
        let mut contents = String::new();
        file.read_to_string(&mut contents)?;
        Ok(contents)
    }
    fn write_string(&self, path: &str, content: &str) -> io::Result<()> {
        let mut file = std::fs::OpenOptions::new()
            .write(true)
            .create(true)
            .truncate(true)
            .open(path)?;
        file.write_all(content.as_bytes())?;
        Ok(())
    }
}

pub fn format<F: FileSystem>(
    fs: &F,
    path: &str,
) -> Result<(), Box<dyn std::error::Error + Sync + Send>> {
    let input = fs.read_to_string(path)?;
    let input = input.as_str();
    let (_, schema) = Schema::from_sdl(input).finish().map_err(|e| {
        let stack = format!("parser feedback:\n{}", convert_error(input, e));
        stack
    })?;
    // let parsed_schema =
    //     crate::imports::ImportedSchema::new(fs, path).map_err(|x| format!("{}", x))?;
    // let export_schema = crate::export_schema::ExportSchema::new(&parsed_schema);
    // let _ = crate::validate::validate_schema(&export_schema);
    let formatted_input = schema_to_sdl(&schema);
    fs.write_string(path, &formatted_input)?;
    Ok(())
}

pub fn codegen<F: FileSystem>(fs: &F, path: &str) -> Result<(), PrettyError<String>> {
    let parsed_schema =
        crate::imports::ImportedSchema::new(fs, path).map_err(|x| format!("{}", x))?;
    let export_schema = crate::export_schema::ExportSchema::new(&parsed_schema);
    let validated_schema = crate::validate::validate_schema(&export_schema);
    let flat_validated_schema = crate::flat_schema::FlatValidatedSchema::new(&validated_schema);
    match &parsed_schema
        .schema
        .config
        .unwrap_or(parser::Config {
            import: None,
            export: None,
        })
        .export
    {
        Some(export) => {
            let generated_code =
                codegen_per_language(export, &validated_schema, &flat_validated_schema);
            generated_code
                .into_iter()
                .map(|(code_path, content)| {
                    let abs_code_path = resolve_path(path, &code_path)?;
                    println!(
                        "Schema \"{}\" has been exported to \"{}\"",
                        path, abs_code_path,
                    );
                    fs.write_string(&abs_code_path, &content)
                })
                .collect::<Result<Vec<_>, _>>()
                .map_err(|x| format!("{}", x))?;
        }
        None => {
            println!("Schema \"{}\" has no export", path);
        }
    }
    Ok(())
}

enum Command {
    Codegen,
    Format,
}

const VERSION: &str = env!("CARGO_PKG_VERSION");

fn main() -> Result<(), PrettyError<String>> {
    let args: Vec<String> = std::env::args().collect();
    if args[1] == "--version" {
        println!("v{}", VERSION);
        return Ok(());
    }
    if args.len() < 3 {
        return Err("Usage: [codegen/format] <path_to_sdl_file>"
            .to_string()
            .into());
    }
    let command = match args[1].as_str() {
        "format" => Ok(Command::Format),
        "codegen" => Ok(Command::Codegen),
        x => Err(format!("Unknown command \"{}\"", x)),
    }?;

    let file_paths = &args[2..];
    let abs_file_paths = file_paths
        .iter()
        .map(|file_path| {
            let path = PathBuf::from(file_path)
                .canonicalize()
                .map_err(|_| "Couldn't get abs path for schema".to_string())?;
            let abs_file_path = path
                .to_str()
                .ok_or("Couldn't get abs path for schema".to_string())?
                .to_string();
            Ok(abs_file_path)
        })
        .collect::<Result<Vec<_>, PrettyError<String>>>()?;

    let fs = RealFileSystem {};

    let _ = abs_file_paths
        .into_iter()
        .map(|abs_file_path| {
            match command {
                Command::Format => format(&fs, &abs_file_path).map_err(|e| e.to_string())?,
                Command::Codegen => codegen(&fs, &abs_file_path)?,
            }
            Ok(())
        })
        .collect::<Result<Vec<()>, PrettyError<String>>>()?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::{cell::RefCell, collections::HashMap, io};

    use crate::{codegen, format, FileSystem};

    pub struct MockFileSystem {
        pub files: RefCell<HashMap<String, String>>,
    }

    impl FileSystem for MockFileSystem {
        fn read_to_string(&self, path: &str) -> io::Result<String> {
            self.files.borrow().get(path).cloned().ok_or_else(|| {
                io::Error::new(
                    io::ErrorKind::NotFound,
                    format!("File \"{}\" not found", path),
                )
            })
        }

        fn write_string(&self, path: &str, content: &str) -> io::Result<()> {
            self.files
                .borrow_mut()
                .insert(path.to_string(), content.to_string());
            Ok(())
        }
    }

    #[test]
    fn test_parse_example_sdl() -> Result<(), Box<dyn std::error::Error + Sync + Send>> {
        let mock_fs = MockFileSystem {
            files: RefCell::new(HashMap::from([
                (
                    "output/another-schema.memorix".to_string(),
                    r#"
Config {
  export: {
    engine: Redis( env(REDIS_URL) )
    files: [
      {
        language: typescript
        path: "memorix.generated.ts"
      }
      {
        language: python
        path: "memorix_generated.py"
      }
      {
        language: rust
        path: "memorix_generated.rs"
      }
    ]
  }
}
Type { abc: u32 }
Cache {
  number_of_messages: {
    payload: u32
    public: [get]
  }
  messages: {
    key: string
    payload: string
  }
}
"#
                    .to_string(),
                ),
                (
                    "output/schema.memorix".to_string(),
                    r#"
Config {
  import: [
    "another-schema.memorix"
  ]
  export: {
    engine: Redis( env(REDIS_URL) )
    files: [
      {
        language: typescript
        path: "memorix.generated.ts"
      }
      {
        language: python
        path: "memorix_generated.py"
      }
      {
        language: rust
        path: "memorix_generated.rs"
      }
    ]
  }
}

Namespace MessageService {
    Enum {
      Operation {
        START
        STOP
      }
    }
    Cache {
        message: {
            key: string
            payload: {
            id: string
            sender_id: u64
            recipient_id: u64
            content: string
            timestamp: u64
            }
            public: [get]
        }
    }

    PubSub {
        new_message: {
            key: u64
            payload: {
            message_id: string
            recipient_id: u64
            }
            public: [subscribe]
        }
    }

    Task {
        message_processing_tasks: {
            payload: {
                message_id: string
                processing_type: string
                priority: u32
            }
            queue_type: "Fifo"
        }
    }
}

Cache {
  user_profile: {
    key: u32
    payload: {
      id: u32
      name: string
      email: string
    }
    public: [get]
  }
  user_session: {
    key: string
    payload: {
      user_id: u32
      session_token: string
      expiry: u64
    }
    ttl: env(USER_SESSION_TTL)
  }
}
  
Namespace UserService {
  Type {
    UserId: u64
  }

  Cache {
    user_profile: {
      key: UserId
      payload: {
        id: UserId
        name: string
        email: string
      }
      public: [get]
    }
    user_session: {
      key: string
      payload: {
        user_id: UserId
        session_token: string
        expiry: u64
      }
      ttl: env(USER_SESSION_TTL)
    }
  }

  PubSub {
    user_activity: {
      key: UserId
      payload: {
        user_id: UserId
        action: string?
        timestamp: u64
      }
      public: [subscribe]
    }
  }

  Task {
    user_registration_tasks: {
      payload: {
        user_id: UserId
        email: string
        registration_date: u64
      }
      public: [enqueue get_len]
    }
  }

}

"#
                    .to_string(),
                ),
            ])),
        };
        codegen(&mock_fs, "output/schema.memorix")?;
        let typescript_path = "output/memorix.generated.ts";
        let typescript_code = mock_fs.read_to_string(typescript_path)?;
        insta::assert_snapshot!(typescript_code);
        let python_path = "output/memorix_generated.py";
        let python_code = mock_fs.read_to_string(python_path)?;
        insta::assert_snapshot!(python_code);
        let rust_path = "output/memorix_generated.rs";
        let rust_code = mock_fs.read_to_string(rust_path)?;
        insta::assert_snapshot!(rust_code);
        Ok(())
    }

    #[test]
    fn test_fail_nicely() -> Result<(), Box<dyn std::error::Error>> {
        let mock_fs = MockFileSystem {
            files: RefCell::new(HashMap::from([(
                "schema.memorix".to_string(),
                r#"
Config {
  export: {
    engine: Redis(env(REDIS_URL))
    files: [      {
        language: typescript
        path: "memorix.generated.ts"
      }]
  }
}

Type {
  a: [u64 u32]
}

"#
                .to_string(),
            )])),
        };

        let result = codegen(&mock_fs, "schema.memorix");
        let err = result.unwrap_err();
        insta::assert_snapshot!(err.to_string());

        Ok(())
    }
    #[test]
    fn test_format() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let mock_fs = MockFileSystem {
            files: RefCell::new(HashMap::from([(
                "schema.memorix".to_string(),
                r#"
    Config{
      export: {
        engine: Redis(env(REDIS_URL))
        files: [      {
            language: typescript
            path: "memorix.generated.ts"
          }]
      }
    }
    Namespace Science {
      Cache {
        how_many_atoms: {
          payload: u32
          key: u64
          ttl:   env(  TTL_HOW_MANY_ATOMS  )
        }
      }
      PubSub {
        how_many_atoms: {
          payload: {id : u32 number: u64}
        }
        how_many_atoms_2: {
          payload: [{id : u32 number: u64}]
        }
      }
    }
    Namespace Rocket {
      Cache {
        launched: {
          payload: boolean
        }
      }
      Enum {
        Operation {
          START
          STOP
        }
        Color {
          RED
          GREEN
          BLUE
        }
      }
      Namespace aa {
        Cache {
          bb: {
            payload:boolean
          }
        }
      }
      PubSub {
        launched: {
          payload: boolean   ?
        }
      }
    }
    Type {
      a: [u64      ]
    }
    "#
                .to_string(),
            )])),
        };
        format(&mock_fs, "schema.memorix")?;
        let formatted_schema = mock_fs.read_to_string("schema.memorix")?;
        insta::assert_snapshot!(formatted_schema);

        Ok(())
    }
}
