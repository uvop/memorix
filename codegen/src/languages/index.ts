import { assertUnreachable } from "src/utilities";
import { codegenTs } from "./ts";
import { codegenPython } from "./python";

export enum Languages {
  typescript,
  python,
}

type CodegenByLanguageFn = (schema: string, language: Languages) => string;

export const codegenByLanguage: CodegenByLanguageFn = (scope, language) => {
  switch (language) {
    case Languages.typescript:
      return codegenTs(scope);
    case Languages.python:
      return codegenPython(scope);
    default:
      assertUnreachable(language);
      return "";
  }
};
