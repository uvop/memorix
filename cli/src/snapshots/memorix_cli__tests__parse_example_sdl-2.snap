---
source: src/main.rs
assertion_line: 423
expression: python_code
---
# flake8: noqa
import typing
import os

if typing.TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from memorix_client_redis import dataclass

from enum import Enum
from memorix_client_redis import (
    MemorixBase,
    MemorixCacheBase,
    MemorixCacheItem,
    MemorixCacheItemNoKey,
    MemorixPubSubBase,
    MemorixPubSubItem,
    MemorixPubSubItemNoKey,
    MemorixTaskBase,
    MemorixTaskItem,
    MemorixTaskItemNoKey,
)


@dataclass
class InlineCachePayloadUserProfile(object):
    id: int
    name: str
    email: str


@dataclass
class InlineCachePayloadUserSession(object):
    user_id: int
    session_token: str
    expiry: int

abc = int

class MessageService(object):

    class Operation(str, Enum):
        START = "START"
        STOP = "STOP"


    @dataclass
    class InlineCachePayloadMessage(object):
        id: str
        sender_id: int
        recipient_id: int
        content: str
        timestamp: int


    @dataclass
    class InlinePubSubPayloadNewMessage(object):
        message_id: str
        recipient_id: int


    @dataclass
    class InlineTaskPayloadMessageProcessingTasks(object):
        message_id: str
        processing_type: str
        priority: int


    class MemorixCache(MemorixCacheBase):
        def __init__(self, api: MemorixBase) -> None:
            super().__init__(api=api)

            self.message = MemorixCacheItem[str, MessageService.InlineCachePayloadMessage, True, True, True](
                api=api,
                id="message",
                payload_class=MessageService.InlineCachePayloadMessage,
            )


    class MemorixPubSub(MemorixPubSubBase):
        def __init__(self, api: MemorixBase) -> None:
            super().__init__(api=api)

            self.new_message = MemorixPubSubItem[int, MessageService.InlinePubSubPayloadNewMessage, True, True](
                api=api,
                id="new_message",
                payload_class=MessageService.InlinePubSubPayloadNewMessage,
            )


    class MemorixTask(MemorixTaskBase):
        def __init__(self, api: MemorixBase) -> None:
            super().__init__(api=api)

            self.message_processing_tasks = MemorixTaskItemNoKey[MessageService.InlineTaskPayloadMessageProcessingTasks, True, True, True, True](
                api=api,
                id="message_processing_tasks",
                payload_class=MessageService.InlineTaskPayloadMessageProcessingTasks,
                options=MemorixTaskItem.Options(
                    queue_type='Fifo',
                )
            )

    class Memorix(MemorixBase):
        def __init__(self, ref: MemorixBase) -> None:
            super().__init__(ref=ref)


            self._namespace_name_tree = ["MessageService"]
            self.cache = MessageService.MemorixCache(self)
            self.pubsub = MessageService.MemorixPubSub(self)
            self.task = MessageService.MemorixTask(self)


class MemorixCache(MemorixCacheBase):
    def __init__(self, api: MemorixBase) -> None:
        super().__init__(api=api)

        self.user_profile = MemorixCacheItem[int, InlineCachePayloadUserProfile, True, True, True](
            api=api,
            id="user_profile",
            payload_class=InlineCachePayloadUserProfile,
        )
        self.user_session = MemorixCacheItem[str, InlineCachePayloadUserSession, True, True, True](
            api=api,
            id="user_session",
            payload_class=InlineCachePayloadUserSession,
            options=MemorixCacheItem.Options(
                ttl=os.environ['USER_SESSION_TTL'],
            )
        )
        self.number_of_messages = MemorixCacheItemNoKey[int, True, False, False](
            api=api,
            id="number_of_messages",
            payload_class=int,
        )

class Memorix(MemorixBase):
    def __init__(self) -> None:
        super().__init__(redis_url=os.environ['REDIS_URL'])


        self._namespace_name_tree = []
        self.MessageService = MessageService.Memorix(ref=self)

        self.cache = MemorixCache(self)
